### ISSUE TITLE
sending a message is often delayed, sometimes buggy

### ISSUE THREAD

**Issue by hpk42** [deltachat/deltachat-desktop#3586]
am using a recent master of core and desktop. 

when i hit "ctrl-enter" for sending -- half the time it delays for 2+ seconds and sometimes the draft-text does not go away even after the message is posted (see screenshot).  It feels like something is blocking.  Currently, i can't offer much help in debugging but at least wanted to raise it, anyway. 

![Screenshot from 2023-12-07 13-36-11](https://github.com/deltachat/deltachat-desktop/assets/73579/fc4a298b-9db0-4daf-a1bb-82116de20c48)

**Comment by Simon-Laux** [comment]
maybe send me the log after it happened, maybe there is some error somewhere?

**Comment by WofWca** [comment]
This code is responsible for message sending and draft removal:

https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/packages/frontend/src/components/composer/Composer.tsx#L140-L153

It calls `__reloadDraft`, which is where the input text is reset, I believe:

https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/packages/frontend/src/components/composer/Composer.tsx#L448

**Comment by hpk42** [comment]
Can't comment on the code but the issues still appears often for me -- on desktop i now take care especially in group chats, to very carefully check if the message went out, even if the input field continues to have the full draft message.  I then need to remove it manually.

**Comment by WofWca** [comment]
Is it bad on all your Delta Chat accounts?

**Comment by hpk42** [comment]
On Thu, Sep 19, 2024 at 23:59 -0700, WofWca wrote:
> Is it bad on all your Delta Chat accounts?

Only my (big) main accounts exposes the problem, and only in bigger group chats. 
In other accounts i am not really part of larger chats. 
Superficially, this points to some event-bubbling/timing problem? 
I also never had this problem on Android.

**Comment by WofWca** [comment]
@Simon-Laux could you explain what's the point of `__reloadDraft()` instead of `clearDraft()` https://github.com/deltachat/deltachat-desktop/pull/2529 ?

**Comment by Simon-Laux** [comment]
> @Simon-Laux could you explain what's the point of __reloadDraft() instead of clearDraft() https://github.com/deltachat/deltachat-desktop/pull/2529 ?

to show the actual state of the draft.

**Comment by WofWca** [comment]
But I mean, can it so happen that there is gonna be a draft after sending a message? Why make an async call. Would `clearDraft()` + `__reloadDraft()` make sense at least?

**Comment by WofWca** [comment]
@hpk42 A fix for a related issue (https://github.com/deltachat/deltachat-core-rust/issues/6036) has been released with Delta Chat Dekstop 1.47.1. Please report whether things improved.

**Comment by hpk42** [comment]
K, i'll close and will re-open or open a new issue if it re-appears. 
thanks!

**Comment by WofWca** [comment]
Hmmm I think I just encountered this issue (draft not getting removed) myself, on 1.49.0 (git: v1.33.0-1229-gc574dd42)

Could have to do with #4144 where we switched to `throttle` to save draft.

I suspect this mostly happens on heavier accounts.

**Comment by hpk42** [comment]
yesterday, on desktop-main the draft did not disappear often,
while the message was sent out. The draft was never cleared. 
This happened around 5 out of 6 times when i sent a message. 
Today it doesn't happen so much -- but maybe today is also less chat activity
on my account, not sure. 

Also regarding other problems with message sending
(eg on android https://github.com/deltachat/deltachat-android/issues/3489 )
i am wondering if there should be a more direct "send draft message" API
so that the UI executes an atomic action on core, that adds/sends the message
and clears the draft in one go.

**Comment by Simon-Laux** [comment]
> i am wondering if there should be a more direct "send draft message" API

yes, that is actually the main idea of the composer rewrite. keep more of the state in core and don't recreate the draft every time it is changed.

**Comment by r10s** [comment]
> i am wondering if there should be a more direct "send draft message" API
so that the UI executes an atomic action on core, that adds/sends the message
and clears the draft in one go.

this is already in place since since https://github.com/deltachat/deltachat-core-rust/pull/2887 - you can send_msg(draft_id) that sends the message out and clears the draft.

there was also https://github.com/deltachat/deltachat-core-rust/pull/4839 added a long time ago, to make webxdc-staging work.

so, from view of core and jsonrpc, things are fine - it is just that things are not (always) used that way

**Comment by nicodh** [comment]
Is this still an issue?

**Comment by WofWca** [comment]
After https://github.com/deltachat/deltachat-desktop/pull/5049 I'm not seeing how else this can be still happening besides `miscSetDraft` being already in flight when [`composerSendMessage`](https://github.com/deltachat/deltachat-desktop/blob/264319c0d99857731396cc493a2db27915f9fd41/packages/frontend/src/components/composer/Composer.tsx#L202-L278) gets executed.

There was a bit of a discussion about the recently added [`rpc.miscSendDraft`](https://github.com/chatmail/core/blob/fc81cef113c108fda1ef0eddfc761e52601c06cc/deltachat-jsonrpc/src/api.rs#L2503-L2515) being the supposed solution, but I'm skeptical of the usefulness of such a function, because it requires that you wait for the draft to get saved before you can send the message. Which is a bit of a headache, since you have to manage the possibility that the user has typed more after the call to `miscSetDraft` but before it has resolved. What sounds better to me is something like `sendMessageAndRemoveDraft` (or changing the current [`misc_send_msg`](https://github.com/chatmail/core/blob/fc81cef113c108fda1ef0eddfc761e52601c06cc/deltachat-jsonrpc/src/api.rs#L2381-L2455) to do that).

Anyways, here are 3 more things that can solve or minimize the issue:
- only save the draft when necessary (see also https://github.com/deltachat/deltachat-desktop/issues/5252), i.e. on big debounce, or when switching chats. This minimizes the amount of `miscSetDraft` calls, thus reducing the probability of a race.
- prior to `miscSendMsg`, ensure that there are no pending `miscSetDraft` calls. `await` the pending one(s). But this has the same problems as with `miscSendDraft` (see above).
- As mentioned above, add `sendMessageAndRemoveDraft`. From the API standpoint this seems like the most reasonable approach. It's just not a thing that you want to send a message, but keep another draft for this chat.
